torch/csrc/jit/passes/create_autodiff_subgraphs.cpp

```diff
@@ -12,6 +13,66 @@ namespace jit {
 
 namespace {
 
+std::vector<c10::optional<const Use>> gatherLastUses(
+    at::ArrayRef<Value*> values) {
+  return fmap(values, [&](Value* v) -> c10::optional<const Use> {
+    return firstOrLastUse(v, /*find_first*/ false);
+  });
+}
+
+// When merging a node into a subgraph, we wish to preserve all of the
+// aliasing properties of the node's outputs. It is difficult to track
+// the node or its contained nodes through all of the ir manipulation
+// involved in merging; it is pretty easy to uniquely identify the value
+// based on its uses. We can identify the value by its last use in the graph.
+// Values which do not have uses or which do not have a last use
+// outside of the subgraph to be merged into we do not need to track.
+struct ValueMapper {
+  ValueMapper(Node* n, AliasDb& db, size_t subgraph_num_outputs) {
+    last_uses_ = gatherLastUses(n->outputs());
+    subgraph_num_outputs_ = subgraph_num_outputs;
+    WithInsertPoint guard(n);
+    auto g = n->owningGraph();
+    // temporary node to put the aliasing properties of the node before its
+    // merged and destroyed
+    placeholder_node_ = g->insertNode(g->create(prim::Uninitialized, 0));
+    for (size_t i = 0; i < n->outputs().size(); ++i) {
+      Value* existing = n->outputs().at(i);
+      Value* new_value =
+          placeholder_node_->insertOutput(i)->copyMetadata(n->outputs().at(i));
+      db.replaceWithNewValue(existing, new_value);
+    }
+  }
+
+  bool usesEqual(const Use& a, const Use& b) {
+    return a.user == b.user && a.offset == b.offset;
+  }
+
+  void copyAliasing(Node* merged_node, AliasDb& db) {
```

Comment:
nitpick: should we just store `AliasDb` as a member of `ValueMapper` since we are already passing it in a c-tor which simplifies `copyAliasing` a little bit?